updating `set_frequency` to include fine tune frequency

alexrkaufman opened the pull request:
Moving the conversation from #7 about the `set_frequency` function into it's own thread for discussion. 

> https://github.com/alexrkaufman/pytla/pull/7#issuecomment-2233513393
> re: `set_frequency` function changes
> 
> I like the goal of consolidating setting frequency into a single function.
> 
> From what I recall, and forgive me if I am incorrect because it has been a hot minute since I had access to an ITLA, the fine tune frequency register does not reset to zero when you set the frequency registers. This could cause confusion for a user where someone sets the frequency but finds that it the frequency is offset without a clear reason because the fine tune frequency was set previously. This is why the old `set_frequency` function reset the fine tune frequency to zero each time it was called.
> 
> Maybe something like this would work? The problem I see here is that it does provide for an option that leaves the fine tune frequency alone.
> 
> ```python
> def set_frequency(self, freq, ftf=0):
> 
> ... the rest of the function ...
> 
>     try:
>         self.set_fine_tuning(ftf)
>     except CPException:
>         self.wait()
>     except ExecutionError:
>         self.nop()
> ```



@f3fora you are probably working more closely with one of these lasers than I am now. I will defer to your judgement about what a typical ITLA user would expect the behavior of `set_frequency` to be. 

The goal of this project was to streamline a lot of the difficulties working with registers directly. I want to make sure that users are not surprised by the behavior of functions. I don't necessarily think how I originally set this up achieves that goal. Let me know what you think.

Commit:
feat(set_frequency): change only first frequency and disable ftf

Commit:
feat(ftfr): introduce fine-tuning range

f3fora commented:
At the end, I am using this modify version of your original one, in which I solve the problem of some execution error.

One should note, that the laser should be disabled.

```py
    def set_frequency(self, freq):
        """Sets the frequency of the laser in TeraHertz.

        Has MHz resolution. Will round down.

        This function sets the first channel frequency and then sets the
        channel to 1 so that the frequency output when the laser is enabled
        will be the frequency given to this function.

        If you would like to only change the first channel frequency use
        the function `set_fcf`.

        Disable the laser before calling this function.

        :param freq: The desired frequency setting in THz.
        :returns: None
        """

        # This does a check so this only runs if fine tuning has been turned on.
        if self.get_fine_tuning() != 0:
            # Set the fine tuning off!
            while True:
                try:
                    self.set_fine_tuning(0)
                except ExecutionError:
                    sleep(self.default_sleep_time)
                except CPException:
                    self.wait()
                    break
                else:
                    break

        try:
            self.set_fcf(freq)
        except CPException:
            self.wait()

        try:
            self.set_channel(1)
        except CPException:
            self.wait()

```


I have also try something more exotic. This will allow to keep the laser on. but it is extremely slow.

```py
    def get_ftf_range(self):
        response = self._ftfr()

        ftfr = int.from_bytes(response, "big")

        return ftfr * 1e-3

    def set_frequency_on(self, freq):
        # get the fine-tuning range
        ftfr_GHz = self.get_ftf_range()
        # get the grid spacing
        grid_GHz = self.get_grid()

        freq_GHz = freq * 1e3
        # get the current grid position
        current_fcf_GHz = self.get_fcf() * 1e3

        # compute the position on the grid that is the closest to the wanted frequency
        diff_GHz = freq_GHz - current_fcf_GHz
        n = int(np.round(diff_GHz / grid_GHz))
        fcf_GHz = current_fcf_GHz + n * grid_GHz
        # the difference between the position on the grid and the
        finetune_GHz = freq_GHz - fcf_GHz

        print(fcf_GHz * 1e-3, finetune_GHz)
        assert finetune_GHz < ftfr_GHz

        if n < 0:
            raise Exception("Disable Laser and Change FCF freq!")

        while True:
            try:
                self.set_fine_tuning(finetune_GHz)
            except ExecutionError:
                time.sleep(self.default_sleep_time)
            except CPException:
                self.wait()
                break
            else:
                break

        if n > 0:
            try:
                self.set_channel(n + 1)
            except CPException:
                self.wait()

```

alexrkaufman commented:
@f3fora I think this is good to go. for setting the frequency and the `get_ftfr_range` function. 

Can you double check that the ITLA1.3 `set_frequency` function looks correct. When I rebased onto main and removed my original commit it got a little confusing fixing the conflicts. Just want to ensure I dont merge something totally borked. With your 👍 I think this is good to go.

alexrkaufman committed:
fixing sleep_time typos in ITLA12 and ITLA13

self.default_sleep_time --> self.sleep_time

f3fora commented:
Look OK!

alexrkaufman commented:
@f3fora One note here on your comment about the `set_frequency_on` function.

 If I understand what that function is doing it can only change the frequency while the laser is on within the fine tuning range. It may be beneficial to include checks that ensure fine tuning changes are within the fine tuning range to `set_fine_tuning`. The other functionality may be better implemented downstream by the projects using the library, although this is a softly held opinion and I could be convinced otherwise. 

Also, thank you for picking this project up! I built this because I needed it in grad school and have always hoped people have found it useful. For you to take it up and push it forward with fixes means a lot.

f3fora commented:
> @f3fora One note here on your comment about the `set_frequency_on` function.
> 
> If I understand what that function is doing it can only change the frequency while the laser is on within the fine tuning range. It may be beneficial to include checks that ensure fine tuning changes are within the fine tuning range to `set_fine_tuning`. The other functionality may be better implemented downstream by the projects using the library, although this is a softly held opinion and I could be convinced otherwise.

The frequency `f` of the laser  (from OIF ITLA MSI specification - 9.6.7) is given by :
```
f = fcf + (no_channel - 1) * grid_spacing + ftf
```
where `fcf` is the first channel frequency, `no_channel` is the number of channel, `grid_spacing` is the spacing between the channel (as frequency) and `ftf` is the fine tuning frequency.

Given that `no_channel` and `ftf` can be changed without turning off the laser, I set the frequency in this way.It turn out to be really slowly, since the ftf adjust is slow.

Moreover, the range of `ftf` is [-`ftfr` , +`ftfr`] and given that 2 * `ftfr` always > `grid_spacing` the fine tune frequency that I set is always in the range.

The current implementation of `set_frequency` is faster (at least in my setup), but requires to turn off and on the laser. 
Moreover it is required to set the `ftf` to zero to achieve the right frequency (from OIF ITLA MSI specification 9.8.7)

> If the FTF command is issued when the laser is not enabled (SENA = 0), the laser frequency will be set to the sum of the channel frequency and the FTF frequency when the laser output is enabled

Commit `71e873851bea61e0bec4e1f468d35a85cce599fd` should thus be reverted or one should subtract (get_fine_tuning) to the value passed to `set_fcf`, e.g.

```py
    def set_frequency(self, freq, reset_ftf=True):
        """Sets the frequency of the laser in TeraHertz.

        Has MHz resolution. Will round down.

        This function sets the first channel frequency and then sets the
        channel to 1 so that the frequency output when the laser is enabled
        will be the frequency given to this function.

        If you would like to only change the first channel frequency use
        the function `set_fcf`.

        Disable the laser before calling this function.

        :param freq: The desired frequency setting in THz.
        :param reset_ftf: boolean if the user wants to reset the fine tune frequency.
        :returns: None
        """

        # This does a check so this only runs if fine tuning has been turned on.
        fine_tuning =  self.get_fine_tuning() 
        if reset_ftf and fine_tuning != 0:
            # Set the fine tuning off!
            while True:
                try:
                    self.set_fine_tuning(0)
                except ExecutionError:
                    sleep(self.sleep_time)
                except CPException:
                    self.wait()
                    fine_tuning = 0
                    break
                else:
                    fine_tuning = 0
                    break

        try:
            self.set_fcf(freq - fine_tuning)
        except CPException:
            self.wait()

        try:
            self.set_channel(1)
        except CPException:
            self.wait()
```

> Also, thank you for picking this project up! I built this because I needed it in grad school and have always hoped people have found it useful. For you to take it up and push it forward with fixes means a lot.

Thanks for your work!

alexrkaufman commented:
Reverted commit 71e873. Thank you for identifying that adding that option caused a discrepancy between the library and the MSA.
